clean_air/util/cubes.py

```diff
@@ -18,3 +22,107 @@ def get_xy_coords(cube):
     xcoord = cube.coord(axis="x", dim_coords=True)
     ycoord = cube.coord(axis="y", dim_coords=True)
     return xcoord, ycoord
+
+
+def extract_box(cube, box):
+    """
+    Extracts a rectangular area from a cube.
+
+    Args:
+        cube: Cube to subset
+        box (x0, y0, x1, y1): Box defined in terms of its lower-left (x0, y0)
+            and upper-right (x1, y1) corners
+    """
+
+    def extent_checker(low, high):
+        """
+        Create a callback function that checks whether a cell is contained
+        in a given range.
+
+        For bounded cells, the cell is included iff its bounded region
+        has nonempty intersection with the requested interval. Otherwise
+        falls back to simply checking the point.
+        """
+        def cb(cell):
+            if cell.bound:
+                a, b = cell.bound
+                # Sufficient to check a <= high and low <= b.
+                # Note that this *does* cover the potential edge case where
+                # a requested range is entirely contained by a cell, as
+                # this is just a < low < high < b
+                return a <= high and low <= b
+            return low <= cell.point <= high
+        return cb
```

Comment:
I might be being stupid here, but why is there a third return?  Is this something to do with it being a callback function?

Reply:
The first, most indented one is inside an `if` block, The second is not, but is still inside the `cb` function. The third is returning *the function itself*.

This is a slightly tricky and technical thing to do I admit, but very helpful and cool (at least I think so). Remember that every `def` introduces a new scope - a new bag of local variables that *everything* inside the function is allowed to access. Including other functions. Whenever any valid variable name is used, first the innermost scope is checked (ie local to the function itself), then if necessary the scope that contained it, and so on up to the module level. Functions therefore need to remember all the relevant levels of scope, or they would extremely break as soon as you start passing them into or out of other functions - which is perfectly allowed, they're objects like any other. That is, `cb` will *always* be able to access the variables `low` and `high` exactly as they were during that particular call to `extent_checker`. We've defined a function and attached some extra state to it. The technical programming term is "closures" if you're interested.

Remember this is replacing `lambda cell: ymin <= cell.point <= ymax`. Technically I could have also done something like
```
lambda cell: cell.bound[0] <= ymax and ymin <= cell.bound[1] if cell.bound else ymin <= cell.point <= ymax
```
but this is clearly a bit unweildy. And then almost exactly the same thing happens later but with `xmin` and `xmax`, so we'd need to remember to apply any future changes in both places. And where would the explanatory comments go?

Reply:
Quite.  This is much nicer than a lambda function, I must admit I'm not a fan of them.  Very nice, thank you for that.